File: src/main/java/com/atrilos/slidingWindow/LengthOfLongestSubstring.java
package com.atrilos.slidingWindow;

import java.util.LinkedHashSet;
import java.util.Set;

/**
 * Given a string s, find the length of the longest substring without repeating characters.
 * <p>
 * <p>
 * <p>
 * Example 1:
 * <p>
 * Input: s = "abcabcbb"
 * Output: 3
 * Explanation: The answer is "abc", with the length of 3.
 * Example 2:
 * <p>
 * Input: s = "bbbbb"
 * Output: 1
 * Explanation: The answer is "b", with the length of 1.
 * Example 3:
 * <p>
 * Input: s = "pwwkew"
 * Output: 3
 * Explanation: The answer is "wke", with the length of 3.
 * Notice that the answer must be a substring, "pwke" is a subsequence and not a substring.
 * <p>
 * <p>
 * Constraints:
 * <p>
 * 0 <= s.length <= 5 * 10^4
 * s consists of English letters, digits, symbols and spaces.
 */
public class LengthOfLongestSubstring {

    public static void main(String[] args) {
        System.out.println(lengthOfLongestSubstring("abcabcbb"));
    }

    public static int lengthOfLongestSubstring(String s) {
        if (s.isEmpty()) return 0;

        Set<Character> set = new LinkedHashSet<>();
        int n = s.length();
        int maxLength = 0;
        int left = 0, right = 0;

        while (right < n) {
            if (!set.contains(s.charAt(right))) {
                set.add(s.charAt(right));
                ++right;
                maxLength = Math.max(maxLength, right - left);
            } else {
                set.remove(s.charAt(left));
                ++left;
            }
        }
        return maxLength;
    }
}

File: src/main/java/com/atrilos/slidingWindow/LengthOfLongestSubstring.kt
package com.atrilos.slidingWindow

/**
 * [3](https://leetcode.com/problems/longest-substring-without-repeating-characters/)
 */
fun lengthOfLongestSubstring(s: String): Int {
    val dict = IntArray(1000)
    var l = 0
    var r = 0
    var res = 0
    while (r < s.length) {
        val current = s[r].code
        dict[current]++
        if (dict[current] == 1) {
            res = maxOf(res, r - l + 1)
        } else {
            while (l <= r && dict[current] > 1) {
                dict[s[l].code]--
                l++
            }
        }
        r++
    }

    return res
}